src/backend/distributed/utils/meta_visibility.c

```diff
@@ -0,0 +1,505 @@
+
+	bool dependsOnCitus = false;
+
+	DependentObjectsContext =
+		AllocSetContextCreate(
+			CurrentMemoryContext,
+			"Dependent Objects Context",
+			ALLOCSET_DEFAULT_SIZES);
+
+	MemoryContext oldContext = MemoryContextSwitchTo(DependentObjectsContext);
+
+	ObjectAddress objectAdress = { metaTableId, objectId, 0 };
+
+	switch (metaTableId)
+	{
+		/* meta classes that access their own oid */
+		case ProcedureRelationId:
+		case AccessMethodRelationId:
+		case EventTriggerRelationId:
+		case TriggerRelationId:
+		case TSConfigRelationId:
+		case TSTemplateRelationId:
+		case TSDictionaryRelationId:
+		case LanguageRelationId:
+		case RewriteRelationId:
+		case AttrDefaultRelationId:
+		case NamespaceRelationId:
+		case ConstraintRelationId:
+		case RelationRelationId:
+		{
+			dependsOnCitus = IsCitusDependentObject(objectAdress, NULL);
+			break;
+		}
+
+		case TypeRelationId:
+		{
+			/*
+			 * we do not access only the typoid in pg_type, we also access typreloid
+			 * to see if type's relation depends on citus, so the type does.
+			 * we always have to access typoid because not all types have a valid
+			 * relation oid (e.g. array, enum).
+			 */
+			dependsOnCitus = IsCitusDependedType(objectAdress);
+			break;
+		}
+
+		/* meta classes that access their typeoid */
+		case EnumRelationId:
+		{
+			/*
+			 * we do not directly access the oid in pg_enum,
+			 * because it does not exist in pg_depend, but its type does
+			 */
+			objectAdress.classId = TypeRelationId;
+			dependsOnCitus = IsCitusDependentObject(objectAdress, NULL);
+			break;
+		}
+
+		/* meta classes that access their relation oid */
+		case IndexRelationId:
+		case AttributeRelationId:
+		case SequenceRelationId:
+		case StatisticRelationId:
+		{
+			objectAdress.classId = RelationRelationId;
+			dependsOnCitus = IsCitusDependentObject(objectAdress, NULL);
+			break;
+		}
+
+		/* meta classes that access their procedure oid */
+		case AggregateRelationId:
+		{
+			objectAdress.classId = ProcedureRelationId;
+			dependsOnCitus = IsCitusDependentObject(objectAdress, NULL);
+			break;
+		}
+
+		default:
+		{
+			break;
+		}
+	}
+
+	MemoryContextSwitchTo(oldContext);
+
+	PG_RETURN_BOOL(dependsOnCitus);
+}
+
+
+/*
+ * IsCitusDependentObject returns true if the given object is dependent on any citus object.
+ * It looks for direct and indirect dependencies of the given object because we have
+ * the objects of some metaclasses (pg_attribute, pg_constraint, pg_index, pg_aggregate,
+ * pg_statistic, pg_sequence) which are not directly stored in 'pg_depend' meta table,
+ * so we use current approach instead of the short one right below (only finds direct deps):
+ *
+ * Oid citusExtensionId = get_extension_oid("citus", false);
+ * ObjectAddress extensionObjectAddress = {ExtensionRelationId, citusExtensionId, 0};
+ * return IsObjectAddressOwnedByExtension(&objectAddress, &extensionObjectAddress);
+ */
+bool
+IsCitusDependentObject(ObjectAddress objectAddress, HTAB *dependentObjects)
```

Comment:
did we try `GetAllDependenciesForObject` function? What is missing there? It seems to recursively pull all the dependencies of a given object.

Reply:
It uses the function _RecurseObjectDependencies_ which takes callbacks for expansion, follow and apply. Current implemented callbacks do not pass the postgres tests because current expansion callback should be changed or a new one should be added such that it contains all required pg meta tables. I think that if we use that api with a new expansion callback, we can get rid of the similar logic inside IsCitusDependentObject and IsCitusDependedType in _citus_depended_object_ file.